### PR TITLE
Add API method LGBM_BoosterPredictForMats

### PR DESCRIPTION
**Pull request by mjmckp**
Add API method `LGBM_BoosterPredictForMats` which runs prediction on a data set given as of array of pointers to rows (as opposed to existing method `LGBM_BoosterPredictForMat` which requires data given as contiguous array).

This is added to allow evaluation on training and validation data sets without reformatting of data in LightGBM.Net (C# wrapper).

Also, added a link to LightGBM.NET C# wrapper.

### CODE REVIEW

**Comment by StrikerRUS** [comment]
@mjmckp Sorry for the inconvenience, but please update the branch to pass CI checks.

**Comment by mjmckp** [comment]
The failing test looks unrelated to these changes (some python error), how to trigger the CI run again?

**Comment by StrikerRUS** [comment]
@mjmckp You are absolutely right! Sorry for the inconvenience with failing tests. We fixed CI tests in #2016. Unfortunately, simple re-running tests without fixed from the mentioned PR will not help, please update the branch to pass tests.

File: src/c_api.cpp [hunk 1]
```diff
@@ -1405,6 +1432,22 @@ RowPairFunctionFromDenseMatric(const void* data, int num_row, int num_col, int d
...
+		auto raw_values = inner_function(0);
+		std::vector<std::pair<int, double>> ret;
+		for (int i = 0; i < static_cast<int>(raw_values.size()); ++i) {
+			if (std::fabs(raw_values[i]) > kZeroThreshold || std::isnan(raw_values[i])) {
+				ret.emplace_back(i, raw_values[i]);
+			}
+		}
+		return ret;
+	};
+}
```
Comment (guolinke) [inline src/c_api.cpp]: can you format these indent to two-space?

**Comment by guolinke** [comment]
@StrikerRUS any ideas why the test failed?

**Comment by StrikerRUS** [comment]
@guolinke Yeah! Refer to the conversation started from https://github.com/Microsoft/LightGBM/pull/2061#pullrequestreview-217047994

**Pull request merged**